python/ray/train/_internal/session.py

```diff
@@ -281,17 +292,64 @@ def _report_thread_runner_error(self, block=False):
         except queue.Empty:
             pass
 
+    def _create_checkpoint_file_list(self, checkpoint: Checkpoint):
+        """Create an index of the folder contents
+
+        So we know which files belong to which rank.
+        """
+        root = checkpoint._local_path
+        ckpt_files = []
+        for dir, _, files in os.walk(root):
+            # Strip the root path from the path though, since
+            # we are only interested in the part relative to
+            # the root of this checkpoint.
+            dir = dir[len(root) :]
+            for fn in files:
+                ckpt_files.append(os.path.join(dir, fn))
+        # Write these files into the index file.
+        with open(os.path.join(root, _INDEX_FILE.format(self.world_rank)), "w") as f:
+            for fn in ckpt_files:
+                f.write(f"{fn}\n")
+
+    def remove_uploaded_checkpoint_files(self, checkpoint: Checkpoint):
+        """Get rid of already uploaded large checkpoint files.
+
+        This is so they don't get shipped to the driver node.
+        """
+        root = checkpoint._local_path
+        for f in os.listdir(root):
+            if f.endswith(_INDEX_FILE_EXTENSION):
+                # We will leave the index file in there so local
+                # checkpoint has knowledge about the cloud files.
+                continue
+            fp = os.path.join(root, f)
+            if os.path.isfile(fp):
+                os.unlink(fp)
+            elif os.path.isdir(fp):
+                shutil.rmtree(fp)
+
     def checkpoint(self, checkpoint: Checkpoint):
         """Adds kwargs to the queue to be consumed by main thread.
 
         Also stores the checkpoint in ``self.loaded_checkpoint``.
         """
+        if (
+            checkpoint._local_path
+            and self.checkpoint_upload_from_workers
+            and self.checkpoint_uri
+        ):
+            self._create_checkpoint_file_list(checkpoint)
+            # We want to upload the files directly to cloud storage,
+            # so that they won't need to be shipped to the driver node
+            # via object store.
+            checkpoint.to_uri(self.checkpoint_uri)
+            self.remove_uploaded_checkpoint_files(checkpoint)
```

Comment:
Talked to @Yard1 about this. Here is what's happening with current code:

1. This works for directory-backed checkpoints only. User (or us developers) provide `Checkpoint.from_directory()` to `session.report`. This makes sense because if it fits in memory already, there's technically no need to jump through these hoops.
2. First, upload to cloud with `to_uri`. Note: we need to nail down whose responsibility it is to do syncing. `Syncer` or `Checkpoint`?
3. Next, delete all local paths -- will this result in `checkpoint._local_path` being deleted? Or just an empty folder?
4. Then, still send the checkpoint to the Train driver for the trainable to handle (in code below):

```python
        result = TrainingResult(
            type=TrainingResultType.CHECKPOINT,
            data=checkpoint,
            metadata=metadata,
        )
        self.result_queue.put(result, block=True)
```

5. This is an empty directory at this point?? **Train will hand off this checkpoint to Tune, and it will dump it to checkpoint_0000x and then try uploading to cloud again.**
    <details>User calls session.report(checkpoint) -> _TrainSession.checkpoint -> Driver will pull this TrainingResult from the queue, which loads into memory as bytes, then ships through object store from worker to driver -> **Driver will use `tune.checkpoint_dir()` and `tune.report()` to report an empty checkpoint + rank 0 worker metrics.**/details>

See:

https://github.com/ray-project/ray/blob/dc6fd827b40e987ad35745c4939e6131ae2569f0/python/ray/train/_internal/checkpoint.py#L250-L255

and

https://github.com/ray-project/ray/blob/dc6fd827b40e987ad35745c4939e6131ae2569f0/python/ray/train/data_parallel_trainer.py#L379-L383

6. Tune will handle the checkpoint again. (Ex: deleting a different checkpoint if needed)

### Proposal

Should we pass `None` in the checkpoint `TrainingResult` for Tune to no-op in the case that we already uploaded it?

Other considerations:
- What about `num_to_keep` handling? That is done by Tune and should delete the entire checkpoint_000x folder.

Reply:
ha, this was a design consideration. the directory is not empty at the point, I do still want to send the index files to driver so that we know the local checkpoint folder is partial.

Reply:
num_checkpoints_to_keep is interesting.
I am not a big fan of that feature, and I think the best we can do now is to make sure they don't get enabled at the same time?

Reply:
I think it's very important, otherwise you can easily run out of disk

Reply:
ok, this doesn't really change anything.
if you set num_to_keep, old local checkpoints will get deleted after you accumulate enough of them.
this is now verified with the same unit test.